<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Goal" data-toc-modified-id="Goal-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Goal</a></span><ul class="toc-item"><li><span><a href="#Notes-and-Observations" data-toc-modified-id="Notes-and-Observations-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Notes and Observations</a></span></li></ul></li><li><span><a href="#Separating-out-the-Test-Data" data-toc-modified-id="Separating-out-the-Test-Data-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Separating out the Test Data</a></span><ul class="toc-item"><li><span><a href="#Saving-Test-Data" data-toc-modified-id="Saving-Test-Data-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Saving Test Data</a></span></li></ul></li><li><span><a href="#Side-observation:-Class-imbalance" data-toc-modified-id="Side-observation:-Class-imbalance-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Side observation: Class imbalance</a></span></li><li><span><a href="#Training-Data-(X,-y)-for-Classifier-1:-Control-vs.-sPTD" data-toc-modified-id="Training-Data-(X,-y)-for-Classifier-1:-Control-vs.-sPTD-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Training Data (X, y) for Classifier 1: Control vs. sPTD</a></span><ul class="toc-item"><li><span><a href="#Saving-X-Data-for-Classifier-1" data-toc-modified-id="Saving-X-Data-for-Classifier-1-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Saving X Data for Classifier 1</a></span></li><li><span><a href="#Saving-y-(label)-Data-for-Classifier-1" data-toc-modified-id="Saving-y-(label)-Data-for-Classifier-1-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Saving y (label) Data for Classifier 1</a></span></li></ul></li><li><span><a href="#Training-Data-(X,-y)-for-Classifier-2:-Control-vs.-PPROM" data-toc-modified-id="Training-Data-(X,-y)-for-Classifier-2:-Control-vs.-PPROM-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Training Data (X, y) for Classifier 2: Control vs. PPROM</a></span><ul class="toc-item"><li><span><a href="#Saving-X-Data-for-Classifier-2" data-toc-modified-id="Saving-X-Data-for-Classifier-2-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Saving X Data for Classifier 2</a></span></li><li><span><a href="#Saving-y-(label)-Data-for-Classifier-2" data-toc-modified-id="Saving-y-(label)-Data-for-Classifier-2-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Saving y (label) Data for Classifier 2</a></span></li></ul></li><li><span><a href="#Next-Step:-Selecting-100-unique-genes" data-toc-modified-id="Next-Step:-Selecting-100-unique-genes-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Next Step: Selecting 100 unique genes</a></span></li></ul></div>

# Goal

This notebook is intended to generate two sets of training/test data for each classifier (Control vs. sPTD and Control vs. PPROM). 

The "raw" data I will be using is from `eset_SC2_v20.RData` file. I have already opened it up in RStudio and saved (`write.csv`) the two objects. The first object `esetSC2` has been saved as `preterm_data.csv`, and the second object `anoSC2` has been saved as `preterm_annotation.csv`.

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv("../data/raw/preterm_data.csv", index_col = 0)
print("There are {} genes and {} samples in preterm_data.csv.".format(data.shape[0], data.shape[1]))
data.head()

There are 29459 genes and 739 samples in preterm_data.csv.


GSM1437802  GSM1437801  GSM1437804  GSM1437803  GSM1437806  \
1_at        5.673714    5.657469    5.263798    5.621857    5.746427   
10_at       3.722012    3.027650    3.095365    3.429769    2.938438   
100_at      6.844537    6.979622    7.311993    7.406321    7.155013   
1000_at     4.209495    3.973889    4.126598    3.721685    3.899342   
10000_at    6.915998    6.820816    6.785624    6.665341    6.637787   

          GSM1437805  GSM1437808  GSM1437807  GSM1437810  GSM1437809  ...  \
1_at        5.382171    5.977089    5.771783    5.511752    5.302418  ...   
10_at       3.055553    2.838617    2.977819    2.944549    2.977487  ...   
100_at      6.930386    7.300295    7.156051    7.512933    7.163658  ...   
1000_at     4.073817    3.811101    3.766648    4.029314    4.153578  ...   
10000_at    6.552980    6.695902    6.714650    6.564450    6.640959  ...   

          Tarca_008_P1A08  Tarca_091_P1H07  Tarca_492_P6D02  Tarca_491_P6C02  \
1_at             5.788344         5.543969         5.128486         5.343032   
10_at            2.919084         2.970130         2.714327         3.012835   
100_at           7.328793         7.189998         7.414430         7.559631   
1000_at          3.718119         3.714274         3.902793         3.870281   
10000_at         6.596466         6.693787         7.315201         7.284721   

          Tarca_506_P6B04  Tarca_920_P10H07  Tarca_919_P10G07  \
1_at             5.577095          5.767927          5.982427   
10_at            2.918379          2.973631          3.044613   
100_at           7.511595          7.325456          7.149270   
1000_at          3.860453          4.062084          4.034426   
10000_at         6.636553          6.501599          6.297040   

          Tarca_918_P10F07  Tarca_917_P10E07  Tarca_916_P10D07  
1_at              5.869221          5.871484          5.752629  
10_at             3.321144          3.039177          2.976313  
100_at            7.130817          7.139528          7.203337  
1000_at           4.021482          3.999080          3.980084  
10000_at          6.600034          6.229878          6.284294  

[5 rows x 739 columns]

In [3]:
annotation = pd.read_csv("../data/raw/preterm_annotation.csv", index_col = 0)
print("There are {} samples and {} columns in preterm_annotation.csv.".format(annotation.shape[0], annotation.shape[1]))
annotation.tail(15)

There are 739 samples and 9 columns in preterm_annotation.csv.


SampleID  IndividualID    GA  GADel Group      Set  Train  \
725   Tarca_626_P7B07         22306  26.9    NaN   NaN  PRB_HTA      0   
726   Tarca_643_P7C09         22306  20.9    NaN   NaN  PRB_HTA      0   
727   Tarca_489_P6A02         22463  30.9    NaN   NaN  PRB_HTA      0   
728   Tarca_505_P6A04         22463  24.3    NaN   NaN  PRB_HTA      0   
729   Tarca_529_P6A07         22463  22.1    NaN   NaN  PRB_HTA      0   
730   Tarca_008_P1A08         22546  22.6    NaN   NaN  PRB_HTA      0   
731   Tarca_091_P1H07         22546  17.6    NaN   NaN  PRB_HTA      0   
732   Tarca_492_P6D02         22572  26.1  32.90  sPTD  PRB_HTA      1   
733   Tarca_491_P6C02         22572  23.1  32.90  sPTD  PRB_HTA      1   
734   Tarca_506_P6B04         22572  19.1  32.90  sPTD  PRB_HTA      1   
735  Tarca_920_P10H07         27755  30.0  30.29  sPTD  PRB_HTA      1   
736  Tarca_919_P10G07         27755  28.1  30.29  sPTD  PRB_HTA      1   
737  Tarca_918_P10F07         27755  26.1  30.29  sPTD  PRB_HTA      1   
738  Tarca_917_P10E07         27755  24.0  30.29  sPTD  PRB_HTA      1   
739  Tarca_916_P10D07         27755  21.0  30.29  sPTD  PRB_HTA      1   

    Platform   TTD  
725    HTA20   NaN  
726    HTA20   NaN  
727    HTA20   NaN  
728    HTA20   NaN  
729    HTA20   NaN  
730    HTA20   NaN  
731    HTA20   NaN  
732    HTA20   6.8  
733    HTA20   9.8  
734    HTA20  13.8  
735    HTA20   0.3  
736    HTA20   2.2  
737    HTA20   4.2  
738    HTA20   6.3  
739    HTA20   9.3

## Notes and Observations

1. SampleID is unique, but same IndividualID can appear multiple times. This is because some women have multiple samples at different GA (gestational age). 
2. GADel is the gestational age delivery, but this information is not available for test data. 
3. TTD is simply the weeks from sample to delivery, or GADel - GA. 

I think we need to make predictions for individuals, not by samples. Since the gestational age is given, perhaps we may be able to ensemble predictions by a weighted average depending on the gestational age. 

# Separating out the Test Data

First I will separate out the test samples from `data` and `annotation` dataframes.

In [27]:
train_and_test_annotation_indices = annotation.groupby("Train").groups
train_and_test_annotation_indices

{0: Int64Index([333, 334, 335, 340, 341, 342, 343, 344, 345, 346,
             ...
             722, 723, 724, 725, 726, 727, 728, 729, 730, 731],
            dtype='int64', length=304),
 1: Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
             ...
             715, 716, 732, 733, 734, 735, 736, 737, 738, 739],
            dtype='int64', length=435)}

In [33]:
test_annotation_indices = train_and_test_annotation[0]
test_annotation = annotation.iloc[test_annotation_indices - 1] # Minus 1 is for 0-based counting in Python
assert test_annotation["Train"].sum() == 0
test_annotation.head()

SampleID  IndividualID    GA  GADel Group      Set  Train  \
333  Tarca_313_P4A04         18275  29.1    NaN   NaN  PRB_HTA      0   
334  Tarca_317_P4E04         18275  25.1    NaN   NaN  PRB_HTA      0   
335  Tarca_378_P4B12         18275  23.1    NaN   NaN  PRB_HTA      0   
340  Tarca_855_P8G11         18427  34.3    NaN   NaN  PRB_HTA      0   
341  Tarca_828_P9D08         18427  30.3    NaN   NaN  PRB_HTA      0   

    Platform  TTD  
333    HTA20  NaN  
334    HTA20  NaN  
335    HTA20  NaN  
340    HTA20  NaN  
341    HTA20  NaN

In [34]:
train_annotation_indices = train_and_test_annotation_indices[1]
train_annotation = annotation.iloc[train_annotation_indices - 1] # Minus 1 is for 0-based counting in Python
assert train_annotation["Train"].sum() == 435
train_annotation.head()

SampleID  IndividualID    GA  GADel    Group       Set  Train Platform  \
1  GSM1437802        515122  28.0   40.0  Control  GSE59491      1   HG21ST   
2  GSM1437801        515122  19.0   40.0  Control  GSE59491      1   HG21ST   
3  GSM1437804        810384  28.0   40.0  Control  GSE59491      1   HG21ST   
4  GSM1437803        810384  18.0   40.0  Control  GSE59491      1   HG21ST   
5  GSM1437806        810392  28.0   41.0  Control  GSE59491      1   HG21ST   

    TTD  
1  12.0  
2  21.0  
3  12.0  
4  22.0  
5  13.0

Now let's subset the `data` Dataframe. I am going to tranpose the dataframe because I prefer to work with `samples-by-genes` matrix.

In [38]:
data = data.transpose()
print(data.shape)
data.head()

(739, 29459)


1_at     10_at    100_at   1000_at  10000_at  100009613_at  \
GSM1437802  5.673714  3.722012  6.844537  4.209495  6.915998      4.460464   
GSM1437801  5.657469  3.027650  6.979622  3.973889  6.820816      3.616890   
GSM1437804  5.263798  3.095365  7.311993  4.126598  6.785624      3.937075   
GSM1437803  5.621857  3.429769  7.406321  3.721685  6.665341      3.490133   
GSM1437806  5.746427  2.938438  7.155013  3.899342  6.637787      3.725999   

            100009676_at  10001_at  10002_at  10003_at  ...   9987_at  \
GSM1437802      6.375718  6.892112  4.310852  2.467458  ...  9.460553   
GSM1437801      6.338360  6.773238  4.657736  2.684089  ...  8.849483   
GSM1437804      6.261533  6.876197  4.446078  2.822903  ...  9.307390   
GSM1437803      6.316306  6.744602  4.695354  2.579157  ...  9.324867   
GSM1437806      6.217101  6.821987  4.483252  2.331567  ...  8.783982   

             9988_at   9989_at    999_at   9990_at    9991_at   9992_at  \
GSM1437802  8.828996  8.869525  5.205051  9.939056   9.981687  3.461509   
GSM1437801  8.333915  9.052965  6.010992  9.991151  10.237839  3.016677   
GSM1437804  8.385302  8.787503  5.503401  9.865705  10.098885  3.213477   
GSM1437803  8.510463  8.665411  5.320511  9.816262  10.116891  3.500357   
GSM1437806  8.263779  8.517233  6.345005  9.795215  10.076414  4.018665   

             9993_at   9994_at   9997_at  
GSM1437802  8.927501  7.001237  7.333485  
GSM1437801  8.816772  6.679151  7.473224  
GSM1437804  8.728658  6.770355  6.917774  
GSM1437803  8.630333  6.660580  7.043207  
GSM1437806  8.810124  6.498137  7.422586  

[5 rows x 29459 columns]

In [47]:
test_SampleIDs = test_annotation["SampleID"]
test_data = data.loc[list(test_SampleIDs)]
print("test_data has {} samples and {} genes.".format(test_data.shape[0], test_data.shape[1]))

test_data has 304 samples and 29459 genes.


In [48]:
train_SampleIDs = train_annotation["SampleID"]
train_data = data.loc[list(train_SampleIDs)]
print("train_data has {} samples and {} genes.".format(train_data.shape[0], train_data.shape[1]))

train_data has 435 samples and 29459 genes.


## Saving Test Data 

Let's save the test data as `X_test.csv` in the `data/processed` folder. I will also save a `torch.Tensor` version in `data/torch` folder to easily access it for inference later.

I will further separate the `train_data` for each of the two classifiers before saving. 

In [49]:
test_data.to_csv("../data/processed/X_test.csv")

In [52]:
import numpy as np
import torch

In [54]:
X_test_np = np.float32(test_data.values)
X_test = torch.from_numpy(X_test_np)
torch.save(X_test, "../data/torch/X_test.pt")

# Side observation: Class imbalance

In [55]:
train_annotation["Group"].value_counts()

Control    285
PPROM       95
sPTD        55
Name: Group, dtype: int64

It was expected but I just noticed the class imbalance in the data. There are far many control samples than PPROM or sPTD samples. If someone knows a good downsampling or upsampling strategies, balancing these numbers may be beneficial depending on the classifier. 

For a first pass, naive classifier, I will not employ these strategies.

Also, for sanity check, there should be 285+55 = **340 total training samples for Control vs. sPTD** and 285+95 = **380 total training samples for Control vs. PPROM** classifiers, respectively.

# Training Data (X, y) for Classifier 1: Control vs. sPTD

I will first extract data for the Control. vs sPTD classifier. 

In [67]:
control_sPTD_annotation = train_annotation[~(train_annotation["Group"] == "PPROM")] # Select all rows not in PPROM group
assert control_sPTD_annotation.shape[0] == 340
control_sPTD_annotation.head()

SampleID  IndividualID    GA  GADel    Group       Set  Train Platform  \
1  GSM1437802        515122  28.0   40.0  Control  GSE59491      1   HG21ST   
2  GSM1437801        515122  19.0   40.0  Control  GSE59491      1   HG21ST   
3  GSM1437804        810384  28.0   40.0  Control  GSE59491      1   HG21ST   
4  GSM1437803        810384  18.0   40.0  Control  GSE59491      1   HG21ST   
5  GSM1437806        810392  28.0   41.0  Control  GSE59491      1   HG21ST   

    TTD  
1  12.0  
2  21.0  
3  12.0  
4  22.0  
5  13.0

In [69]:
control_sPTD_SampleIDs = control_sPTD_annotation["SampleID"]
control_sPTD_data = train_data.loc[list(control_sPTD_SampleIDs)]
print("control_sPTD_data has {} samples and {} genes.".format(control_sPTD_data.shape[0], control_sPTD_data.shape[1]))

control_sPTD_data has 340 samples and 29459 genes.


## Saving X Data for Classifier 1 

Similar to the test data, I will save the training X data as `X_train.csv` in the `data/processed/clf_1` folder. I will also save a `torch.Tensor` for easy access in `data/torch/clf_1` folder.

In [70]:
control_sPTD_data.to_csv("../data/processed/clf_1/X_train.csv")

In [71]:
control_sPTD_np = np.float32(control_sPTD_data.values)
X_control_sPTD = torch.from_numpy(control_sPTD_np)
torch.save(X_control_sPTD, "../data/torch/clf_1/X_train.pt")

## Saving y (label) Data for Classifier 1

I will save the entire control_sPTD_annotation dataframe as `y_train.csv` in `data/processed/clf_1` folder. For `torch.Tensor`, I will first encode the labels as 0 or 1 using `sklearn.preprocessing.LabelEncoder`, then save.

**Note**: The order of the labels is already the same as the rows of the training data! I am not mixing up labels for the samples. :)

Reference to `sklearn.preprocessing.LabelEncoder`: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html

In [72]:
control_sPTD_annotation.to_csv("../data/processed/clf_1/y_train.csv")

In [76]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
le.fit(control_sPTD_annotation["Group"])
print(le.classes_)
le.transform(["Control", "sPTD"])

['Control' 'sPTD']


array([0, 1])

As you can see above, the `Control` has been encoded to class 0 and `sPTD` has been encoded to class 1. 

In [79]:
control_sPTD_y_np = le.transform(control_sPTD_annotation["Group"])
assert control_sPTD_y_np.shape[0] == 340

In [80]:
y_control_sPTD = torch.from_numpy(control_sPTD_y_np)
torch.save(y_control_sPTD, "../data/torch/clf_1/y_train.pt")

# Training Data (X, y) for Classifier 2: Control vs. PPROM

Same as above excpet for the Control. vs PPROM classifier. 

In [81]:
control_PPROM_annotation = train_annotation[~(train_annotation["Group"] == "sPTD")] # Select all rows not in sPTD group
assert control_PPROM_annotation.shape[0] == 380
control_PPROM_annotation.head()

SampleID  IndividualID    GA  GADel    Group       Set  Train Platform  \
1  GSM1437802        515122  28.0   40.0  Control  GSE59491      1   HG21ST   
2  GSM1437801        515122  19.0   40.0  Control  GSE59491      1   HG21ST   
3  GSM1437804        810384  28.0   40.0  Control  GSE59491      1   HG21ST   
4  GSM1437803        810384  18.0   40.0  Control  GSE59491      1   HG21ST   
5  GSM1437806        810392  28.0   41.0  Control  GSE59491      1   HG21ST   

    TTD  
1  12.0  
2  21.0  
3  12.0  
4  22.0  
5  13.0

In [82]:
control_PPROM_SampleIDs = control_PPROM_annotation["SampleID"]
control_PPROM_data = train_data.loc[list(control_PPROM_SampleIDs)]
print("control_sPTD_data has {} samples and {} genes.".format(control_PPROM_data.shape[0], control_PPROM_data.shape[1]))

control_sPTD_data has 380 samples and 29459 genes.


## Saving X Data for Classifier 2

Similar to the test data, I will save the training X data as `X_train.csv` in the `data/processed/clf_2` folder. I will also save a `torch.Tensor` for easy access in `data/torch/clf_2` folder.

In [83]:
control_PPROM_data.to_csv("../data/processed/clf_2/X_train.csv")

In [84]:
control_PPROM_np = np.float32(control_PPROM_data.values)
X_control_PPROM = torch.from_numpy(control_PPROM_np)
torch.save(X_control_PPROM, "../data/torch/clf_2/X_train.pt")

## Saving y (label) Data for Classifier 2

I will save the entire control_sPTD_annotation dataframe as `y_train.csv` in `data/processed/clf_2` folder. For `torch.Tensor`, I will first encode the labels as 0 or 1 using `sklearn.preprocessing.LabelEncoder`, then save.

In [85]:
control_PPROM_annotation.to_csv("../data/processed/clf_2/y_train.csv")

In [87]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
le.fit(control_PPROM_annotation["Group"])
print(le.classes_)
le.transform(["Control", "PPROM"])

['Control' 'PPROM']


array([0, 1])

In [89]:
control_PPROM_y_np = le.transform(control_PPROM_annotation["Group"])
assert control_PPROM_y_np.shape[0] == 380

In [91]:
y_control_PPROM = torch.from_numpy(control_PPROM_y_np)
torch.save(y_control_PPROM, "../data/torch/clf_2/y_train.pt")

# Next Step: Selecting 100 unique genes 